### PR TITLE
Glsl.meta.slang raytracing #3560

### PR DESCRIPTION
**Pull request by jkwak-work**
This PR is only for a code review purpose.

### CODE REVIEW

File: tests/hlsl-intrinsic/shader-execution-reordering/hit-object-reorder-thread.slang [hunk 1]
```diff
@@ -74,14 +76,14 @@ void rayGenerationMain()
         someValues);
 
     uint r = calcValue(hit);
-    // SPIRV: OpReorderThreadWithHitObjectNV
```
Comment (jkwak-work) [inline tests/hlsl-intrinsic/shader-execution-reordering/hit-object-reorder-thread.slang]: Why is this test removed?

File: tests/hlsl-intrinsic/shader-execution-reordering/hit-object-trace-ray.slang.expected
Comment (jkwak-work) [inline tests/hlsl-intrinsic/shader-execution-reordering/hit-object-trace-ray.slang.expected:1]: Why is this file removed?

File: tests/glsl-intrinsic/raytracing/glsl-rayCompute.slang [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                !=
+                rhs[i][j]
+                )
+            {
+                return false;
+            }
+        }
+    }
+    return true;
+}
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/raytracing/glsl-rayCompute.slang:35]: Don't we have a equality testing function already for Matrix?

File: source/slang/slang-ir.cpp [hunk 3]
```diff
@@ -4287,11 +4289,117 @@ namespace Slang
...
+            }
+            else if (kIROp_SPIRVAsmOperandRayCallableFromLocation == op)
+            {
+                varLayoutPointsTo = this->getRayCallableVariableFromLocation(intLitValue);
+            }
+            else
+            {
+                // should be impossible to hit
+                assert(false);
+            }
```
Comment (jkwak-work) [inline source/slang/slang-ir.cpp]: I think this can be better with
```
else
{
    SLANG_ASSERT(kIROp_SPIRVAsmOperandRayCallableFromLocation == op);
    varLayoutPointsTo = this->getRayCallableVariableFromLocation(intLitValue);
}

**Review by jkwak-work — COMMENTED** [review]
It looks good to me overall.

File: tests/hlsl-intrinsic/shader-execution-reordering/hit-object-reorder-thread.slang (hunk 1 above)
Comment (ArielG-NV) [reply tests/hlsl-intrinsic/shader-execution-reordering/hit-object-reorder-thread.slang]: I broke the tests during development; I forgot to add them back since I broke them at the start; I am adding them 👍

File: tests/hlsl-intrinsic/shader-execution-reordering/hit-object-trace-ray.slang.expected
Comment (ArielG-NV) [reply tests/hlsl-intrinsic/shader-execution-reordering/hit-object-trace-ray.slang.expected:1]: these are not tests and go against the .gitignore; they don't do anything

File: tests/glsl-intrinsic/raytracing/glsl-rayCompute.slang (hunk 2 above)
Comment (ArielG-NV) [reply tests/glsl-intrinsic/raytracing/glsl-rayCompute.slang:35]: we do not, other testing code also has an "equals" function for this purpose

File: source/slang/hlsl.meta.slang [hunk 4]
```diff
@@ -9531,22 +9531,32 @@ static const RAY_FLAG RAY_FLAG_SKIP_PROCEDURAL_PRIMITIVES       = 0x200;
 
 __target_intrinsic(hlsl, RayDesc)
 __target_intrinsic(cuda, RayDesc)
+//__target_intrinsic(glsl, RayDesc)
+//__target_intrinsic(spirv, RayDesc)
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang]: Can you either uncomment these or remove?

File: source/slang/hlsl.meta.slang [hunk 5]
```diff
@@ -9822,7 +9832,8 @@ bool __reportIntersection(float tHit, uint hitKind)
 {
     __target_switch
     {
-    case _GL_EXT_ray_tracing: __intrinsic_asm "reportIntersectionEXT";
+    case _GL_EXT_ray_tracing:
+    case glsl: __intrinsic_asm "reportIntersectionEXT";
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang]: Is this change actually needed?
It seems like _GL_EXT_ray_tracing will cover all the GLSL target, doesn't it?

File: source/slang/hlsl.meta.slang [hunk 6]
```diff
@@ -9873,12 +9886,16 @@ void AcceptHitAndEndSearch()
...
 uint3 DispatchRaysIndex()
 {
     __target_switch
     {
-    case hlsl:  __intrinsic_asm "DispatchRaysIndex";
-    case _GL_EXT_ray_tracing: __intrinsic_asm "(gl_LaunchIDEXT)";
+    case hlsl: __intrinsic_asm "DispatchRaysIndex";
+    case _GL_EXT_ray_tracing:
+    case glsl:
+    case glsl:
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang]: This looks to be a mistake.
There should be only one "glsl" case.

File: source/slang/hlsl.meta.slang [hunk 7]
```diff
@@ -11781,7 +12192,70 @@ struct HitObject
...
+    __glsl_extension(GL_NV_shader_invocation_reorder)
+    float GetCurrentTime() {
+        __target_switch
+        {
+        case glsl:
+            __intrinsic_asm "hitObjectGetCurrentTimeNV($0)";
+        case spirv:
+            spirv_asm {
+                result:$$float = OpHitObjectGetCurrentTimeNV &this
+            };
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang]: Shouldn't this have "return"?
```
return spirv_asm { ... };
```

File: source/slang/slang-emit-spirv.cpp [hunk 8]
```diff
@@ -2954,9 +2954,24 @@ struct SPIRVEmitContext
...
+            ensureExtensionDeclaration(UnownedStringSlice("SPV_NV_shader_invocation_reorder"));
+            requireSPIRVCapability(SpvCapabilityShaderInvocationReorderNV);
+            goto rayExtensionAddition;
         case kIROp_VulkanRayPayloadDecoration:
+        case kIROp_VulkanRayPayloadInDecoration:
+            // needed since GLSL will not set optypes accordingly, but will keep the decoration 
+            ensureExtensionDeclaration(UnownedStringSlice("SPV_KHR_ray_query"));
+            requireSPIRVCapability(SpvCapabilityRayQueryKHR);
+            goto rayExtensionAddition;
+rayExtensionAddition:;
```
Comment (jkwak-work) [inline source/slang/slang-emit-spirv.cpp]: Let's not use GOTO within switch statement.
In this case, you can set a temporary variable to be true and handle the action later.
```
bool isRayExtension = false;
...
case kIROp_VulkanCallablePayloadInDecoration:
...
    isRayExtension = true;
    break;
...
if (isRayExtension)
{
   emitOpDecorateLocation(...);
}
````
In general, "goto" should be limited to error handling; not for code reuse.

File: source/slang/slang-ir.cpp [hunk 9]
```diff
@@ -5794,7 +5898,42 @@ namespace Slang
         addInst(i);
         return i;
     }
-
+    IRSPIRVAsmOperand* IRBuilder::emitSPIRVAsmOperandRayPayloadFromLocation(IRInst* inst)
+    {
+        SLANG_ASSERT(as<IRSPIRVAsm>(m_insertLoc.getParent()));
+        const auto i = createInst<IRSPIRVAsmOperand>(
```
Comment (jkwak-work) [inline source/slang/slang-ir.cpp]: It seems that you shouldn't have const here because your return type doesn't have const and "i" will be returned.

File: source/slang/slang-ir.h [hunk 10]
```diff
@@ -2239,6 +2240,8 @@ struct IRAnalysis
     IRDominatorTree* getDominatorTree();
 };
 
+struct DiagnosticSink;
```
Comment (jkwak-work) [inline source/slang/slang-ir.h]: You may need to move this line up to the top where other classes are declared; around line 28~38.

File: source/slang/slang-parser.cpp [hunk 11]
```diff
@@ -7013,6 +7013,27 @@ namespace Slang
         {
             return SPIRVAsmOperand{ SPIRVAsmOperand::NonSemanticDebugPrintfExtSet, parser->ReadToken() };
         }
+        else if (AdvanceIf(parser, "__rayPayloadFromLocation")) {
```
Comment (jkwak-work) [inline source/slang/slang-parser.cpp]: You should keep the coding style same as how the previous code was written.
Please add a new line character between ")" and "{".
I see that this mistake occurs on multiple lines in this PR.

File: source/slang/slang-parser.cpp [hunk 12]
```diff
@@ -7771,6 +7793,20 @@ namespace Slang
             parser->ReadToken(TokenType::Comma);
         }
 
+#define CASE(key, type) if (AdvanceIf(parser, #key)) { auto modifier = parser->astBuilder->create<type>(); \
+    modifier->location = getIntegerLiteralValue(listBuilder.find<GLSLLayoutModifier>()->valToken); listBuilder.add(modifier); } else
```
Comment (jkwak-work) [inline source/slang/slang-parser.cpp]: Is "modifier->location" only for debugging purpose?

**Review by jkwak-work — COMMENTED** [review]
I have some comments.

File: source/slang/hlsl.meta.slang (hunk 5 above)
Comment (ArielG-NV) [reply source/slang/hlsl.meta.slang]: glsl->_GL_EXT_ray_tracing does/did not seem to work, this is needed
I think its a superset->subset type of problem (_GL_EXT_ray_tracing -> glsl is fine, other way is not)

File: source/slang/slang-ir.cpp (hunk 9 above)
Comment (ArielG-NV) [reply source/slang/slang-ir.cpp]: const is used since other functions which do the same thing also set a const ptr for return; likely const is used since in the function we are signaling that the ptr's value won't change value between instantiation and returning

File: source/slang/slang-parser.cpp (hunk 12 above)
Comment (ArielG-NV) [reply source/slang/slang-parser.cpp]: modifier->location is for creating a modifier and tying a glsl layout(location) location value to it

File: source/slang/hlsl.meta.slang (hunk 5 above)
Comment (jkwak-work) [reply source/slang/hlsl.meta.slang]: Do you mean you don't need to have "case _GL_EXT_ray_tracing" when you have "case glsl"?

File: source/slang/slang-ir.cpp (hunk 9 above)
Comment (jkwak-work) [reply source/slang/slang-ir.cpp]: "const" is not a "signal"-ing keyword, and it cannot automatically become non-const.
Something doesn't look right on this code to me.

File: source/slang/hlsl.meta.slang (hunk 5 above)
Comment (ArielG-NV) [reply source/slang/hlsl.meta.slang]: some code & tests enable/use the _GL_EXT_ray_tracing (hlsl->glsl) capability, I use glsl capability instead (but the code is identical since both are glsl)

File: source/slang/slang-ir.cpp (hunk 9 above)
Comment (ArielG-NV) [reply source/slang/slang-ir.cpp]: I did some checking, in godbolt -- it should not compile in MSVC (without a flag), nor does it likely emit different ASM... I will check to see if there is a reason the other functions use a const~

Comment (ArielG-NV) [reply source/slang/slang-ir.cpp]: nope, just a cause of warnings; I suppose I will remove all of the const'ness from these type of functions and if there is a question of why that will be a different discussion

**Comment by jkwak-work** [comment]
Looks good to me
